### ISSUE TITLE
AD cant find route

### ISSUE THREAD

**Issue by raxtor** [Stephan-S/FS19_AutoDrive#745]
I was playing with blue course more deeply and find some strange behaviour. AD is not able find course from specific point but next neighboring point is working correctly.
![fsScreen_2019_10_31_15_37_46](https://user-images.githubusercontent.com/831266/67956820-fd31c480-fbf4-11e9-955f-fd070268add6.png)
![fsScreen_2019_10_31_15_38_06](https://user-images.githubusercontent.com/831266/67956839-01f67880-fbf5-11e9-96e5-de78a8a7926c.png)
Am I doing something wrong?

**Comment by raxtor** [comment]
Check red line direction. For me there isnt any known reason why 1 point should be OK and next closest point (for green course) is ganerating msg about no existing route. AD should be able join to course from every place.

**Comment by Iwan1803** [comment]
Are you sure, that on the green route the machine can reach the goal "Coop L. Fert"?
Put him directly on the green line and in the correct direction and try again.

**Comment by raxtor** [comment]
Rly I dont understand why I should do it? It seems you dont understand that situtation. Pic 1 has issue with pathfinding and wants recalculate network, but it doesnt help. Pic 2 is working OK. Fertilizer was moved about 2m. In addtion this situation is more complicated because Fertilizer is not able find way from field when I use CP and AD for refueling. 
AD should use way A, eventually way B.. imho
![fsScreen_2019_10_31_15_38_06](https://user-images.githubusercontent.com/831266/67965266-06755e00-fc02-11e9-83fc-5a6a32a2aaa7.png)

**Comment by Iwan1803** [comment]
Next time maybe you should describe the whole situation and also translate the message on the screen as i can't do it 😉
So CP is doing the work on the field and AD should drive the way for refilling.
Is there a working way to the refill station on line A or is there only a way on line B?

**Comment by raxtor** [comment]
Sry, yes, when I move Fertilizer to next point (its 2m move) /Pic1 ->Pic2/ than AD is able join to course. But if I move it back to previous point AD again wants recalculate network. Refill station ias possible reach from both direction, no metter if AD will use A or B way

**Comment by Iwan1803** [comment]
On the screen i can't see CP. If CP should start AD, then you need to enter a start and end point in AD, using "Load mode" i think.
You tried a manual AD drive on the screenshots, right?

**Comment by raxtor** [comment]
Sure this situation is result of investigation when I tried localise problem. I know how set up AD and CP. Thats why I use simple driving mode to eliminate CP and AD cooperation isssue.

**Comment by Stephan-S** [comment]
Are you 100% certain there is a path leading to "Coop L. Fert" from F13?

**Comment by raxtor** [comment]
Yes, course for CountyLine 2.0... its near Filed13 if u want test it.
[AutoDrive_County_Line_Seasons_FS19_config.zip](https://github.com/Stephan-S/FS19_AutoDrive/files/3795046/AutoDrive_County_Line_Seasons_FS19_config.zip)

**Comment by Bobster82** [comment]
//thinking out loud, if it helps it helps, if not.... ignore it :P 
@raxtor ,
@Stephan-S , Can't it be the problem with the 90 degrees fix? The vehicle always looks ahead a few waypoints when calculating his route... 

![image](https://user-images.githubusercontent.com/47780758/67987372-84029380-fc2d-11e9-8e9e-42562d1cdec6.png)

**Comment by raxtor** [comment]
I assume that touch point is calculated for closest course than vehicle should join to green course, not to blue and in that case there is no 90 degree problem. Most interesting is when you move 1 point closer to your red arrow, than it works.

**Comment by Bobster82** [comment]
> I assume that touch point is calculated for closest course so than vehicle should join to green course, not to blue and in that case there is no 90 degree problem. Most interesting is when you move vehicle 1 point closer to your red arrow, than it works.

The vehicle isnt always calculating his route from the closest point or line... if the line is facing the other way (to the back) and there is a line nearby (not the closest) thats facing the right way (to the front) its going for the line in the front... If you dont want that, move your vehicle a bit from the line or turn it 180 degrees...

**Comment by raxtor** [comment]
@Bobster82 maybe read whole thread.. AD is not able find path from field. No matter if vehicle is in opposite direction or in correct direction or any other angel to course.

**Comment by Bobster82** [comment]
I did @raxtor (I always do) ,, but I was just popping in with some info, maybe it was something usefull ;)

**Comment by raxtor** [comment]
@Bobster82 but partly u are true that there will problem with 90 degrees limit, thx for navigation.
@Stephan-S cant be it somehow connected with decision which route to use if there is more than 1 route to target? From other side (Field14) it works always. AD is always preferd B route. How role is playing settings "fatest way" for pathfinder in this case? 
![err](https://user-images.githubusercontent.com/831266/68016892-cbbd0580-fc95-11e9-8486-340b243c59e9.png)
Problem is that AD is not able use longer way from this position and laways trying use shortest way...
So question is if route is badly constructed or AD should find way also in this case?

**Comment by Iwan1803** [comment]
Vehicle B can't drive that way as AD doesn't support more than 90° curves.

**Comment by raxtor** [comment]
@Iwan1803 but in that case should check if is not able catch target by longer way, not?

**Comment by Bobster82** [comment]
> Vehicle B can't drive that way as AD doesn't support more than 90° curves.

> @Iwan1803 but in that case should check if is not able reach target by longer way, not?

That was what I was saying... AD takes the shortest route, but it can't find a route cuz its blocked by the 90 degree rule... but its not getting an if its a 90 degree block, try next route (the longer way) 
I dont understand coding that great and I dont know where it is written exactly to check the if statements

**Comment by raxtor** [comment]
@Bobster82 yes, I think we found the problem. Thx.. 
now @Stephan-S should say his opinion how solve it.

**Comment by Iwan1803** [comment]
Why should Stephan solve a non existing error?
It would be much easier if you add a better route 😉

**Comment by Stephan-S** [comment]
Hey,
I do believe that there is an error in the pathfinding/recalculation step.
The 90° degrees angles are discarded during recalculation anyhow, so only the path along the other direction of the blue line would be valid in this case. So I don't think this was the issue here.

But I also noticed an error with blue lines in my savegame.
@raxtor Can you recalculate with the current version? So pack your own zip file from the current repository?
Thank you

**Comment by raxtor** [comment]
@Stephan-S now coursre is finded always, but vehicle is still trying follow short way (orange way) and ofc stucked in circle cuz turn is too sharp (red way). Should continue by green route...
![fsScreen_2019_11_01_14_06_57](https://user-images.githubusercontent.com/831266/68026919-52331080-fcb1-11e9-88cc-abc164f5bdac.png)

**Comment by raxtor** [comment]
@Stephan-S btw I am still thinking about advenced blue course to define routes more comfortable. What do you think about possibility to avoid collision for oncoming vehicles?
![Untitled-1](https://user-images.githubusercontent.com/831266/68029062-b1475400-fcb6-11e9-8459-5ffbd0ea1114.png)

**Comment by raxtor** [comment]
Anyway I am still fan for making blue road with defined left/right offset as main route and green route as helping route for extra tight roads orr other situations. It will simplify construction of road network much more comfortable.

**Comment by raxtor** [comment]
@Stephan-S 
And another question is if we rly need 1 way road at all? Wouldnt be enough just any kind od direction marker?
Than we can have only 1 bi-direction road and two types of route marks:
- direction mark
- offset marker start/end which is defining new offset from dafault settings

**Comment by Stephan-S** [comment]
@raxtor We already had this discussion. I prefer it the way it works now!
Offset for dual way roads is an option but not a priority for me.
It doesn't make the mod any easier for new players and you are forgetting about so many possible problems with your approach.
So, short answer: No :-)

**Comment by raxtor** [comment]
@Stephan-S are you sure? Just imagine that clean network and saved time on big maps.. :-) ok.. but I had to try it.

**Comment by Stephan-S** [comment]
I close this one for now @raxtor :D 
Keep trying and maybe someone will listen to you :-)

**Comment by raxtor** [comment]
@Stephan-S but is it fixed? 1.0.6.6c didnt seem so..

**Comment by raxtor** [comment]
@Stephan-S I tested lates github version and problem still exist...Should I understand you dont plan fix it?

**Comment by raxtor** [comment]
@Stephan-S 90° limit was removed again? I am rly confused now... should I report something yet or should I be completly quiet? Well working unloaders are broken after active unloading, thread unpinned and forgotten. I reported  troubles there nobody replying. Here is closed unsolved mystery. 
Well I will take break...
![fsScreen_2019_11_03_16_06_22](https://user-images.githubusercontent.com/831266/68087279-81ce4e00-fe54-11e9-918e-5c5b83b8da07.png)

**Comment by Stephan-S** [comment]
He will use these routes if there is no other alternative. That looks like there is no path going through the blue line to Coop L. Fert, so he has no other option but to take this turn

**Comment by raxtor** [comment]
If do you think...

**Comment by Stephan-S** [comment]
Ok, so I've spent some time on this issue and I can confirm that it exists and I can also pretty much confirm that I won't / can't fix it with the way AutoDrive currently works.

So just to get this out of the way: I fixed the 90° issue again in my current code base. Will be uploaded soon.

The problem is (which is why I don't like the blue line very much) that the Dijkstra algorithm I use to calculate the fastest paths doesn't care about 'direction' or angles or anything like that.
The 90° limitation is also only a hard work around. 
So if during the calculation all shortest paths to the target AA are calculated.
You would like to get the orange path from start to target but you won't get this.
What you would get with the normal Dijkstra algorithm is the blue line ( C ).
Now my hard work around to prevent 90° angles removes this option in the entry to the blue path.
Now the Dijkstra algorithm cannot produce another shortest path and this will result in not reacheable destination AA from certain points in the network.

So to prevent this. Make sure that there is no path with an angle over 90° that is a shortcut to the target you want to get to.
In this case, this is path C.

For all nodes along the dual way section ( marked with purple D in the image ) the shortest path will always go through C.
It doesn't care about direction.
So if you want to use dual way roads it would be best to create the entries and exits in a way that they are also connected with a green line.
I sketched the fix in dark red.

![Dual way entries](https://user-images.githubusercontent.com/5493976/68089809-f01f0a80-fe6c-11e9-926c-ff018f1c44f7.png)

**Comment by TyKonKet** [comment]
@Stephan-S Why do you have to "lock" 90° angle routes?

**Comment by Stephan-S** [comment]
Otherwise, every driver will see the entry and exit to a dual way route as viable connection and tries to turn almost ~180° in a few meters.

The 90° was just my abstraction of it, to also prevent these sharp turns that most likely won't work as intended everywhere else.

**Comment by TyKonKet** [comment]
@Stephan-S thanks now it's all clear.
I think we should change approach to really fix that kind of problem, maybe turning dual ways routes into two overlapped single way route could help?
Another thing that I think could help is some kind of routes post processing algorithm that smooths routes so that you will never have a "real" turn with more than 10°-15°, this could also reduce number of waypoints on straight ways.
It would also, obviously, improve the automatic driving, since routes would be smoother.

These are just ideas, tell me if I'm wrong or I'm talking bullshit

**Comment by Iwan1803** [comment]
Da kein weiteres Feedback -> geschlossen.
Bitte mit der neuesten Version erneut versuchen und bei Bedarf ein neues Issue erstellen, danke!
_Since no further feedback -> closed.
Please try with the latest version and create a new issue if necessary, thanks!_
